src_firmware/src/driver.c

```diff
@@ -4,296 +4,485 @@
-static uint32_t max_mrpm;// max rpm set by user in init (read from documentation)
-static uint32_t speed_control; // PID output -> pwm calculations input
-
-
-/// ENCODER - variables used for actual speed calculation based on encoder pin and timer interrupts
-static uint64_t count_cycles;
-static uint64_t old_count_cycles;
-static uint64_t count_timer;
-struct k_timer my_timer;
-
-/// BOOLS - drv init and on?
-static bool drv_initialised; // was init command sent?
-static bool is_motor_on; // was motor_on function called?
-
-
-/// PINS definitions
-/// motor out
-static const struct pwm_dt_spec pwm_motor_driver = PWM_DT_SPEC_GET(DT_ALIAS(pwm_drv_ch1));
-static const struct gpio_dt_spec set_dir_p1 = GPIO_DT_SPEC_GET(DT_ALIAS(set_dir_p1_ch1), gpios);
-static const struct gpio_dt_spec set_dir_p2 = GPIO_DT_SPEC_GET(DT_ALIAS(set_dir_p2_ch1), gpios);
-
-/// enc in
-static const struct gpio_dt_spec enc_p1_ch1 = GPIO_DT_SPEC_GET(DT_ALIAS(get_enc_p1_ch1), gpios);
-static const struct gpio_dt_spec enc_p2_ch1 = GPIO_DT_SPEC_GET(DT_ALIAS(get_enc_p2_ch1), gpios);
-static const struct gpio_dt_spec enc_ch1_pins[2] = {enc_p1_ch1, enc_p2_ch1};
-static struct gpio_callback enc_ch1_cb[2];
-
-#if defined(CONFIG_BOARD_NRF52840DONGLE_NRF52840)
-static const struct gpio_dt_spec out_boot = GPIO_DT_SPEC_GET(DT_ALIAS(enter_boot_p), gpios);
-#endif
-
+/// temporary debug only variables: - To be deleted after developemnt is finished!
+uint64_t count_timer;
 static int32_t ret_debug = 100;// DEBUG ONLY
 
-static int speed_pwm_set(uint32_t value)
-{
-	int ret;
-
-	if (!drv_initialised) {
-		return NOT_INITIALISED;
-	}
-
-	if (value > max_mrpm) {
-		return DESIRED_SPEED_TO_HIGH;
-	}
-
-	if (target_speed_mrpm < max_mrpm/10) {
-		value = 0;
-		speed_control = 0;
-	}
-
-	uint64_t w_1 = pwm_motor_driver.period * (uint64_t)value;
-	uint32_t w = value != 0 ? (uint32_t)(w_1/max_mrpm) : 0;
-
-	ret = pwm_set_pulse_dt(&pwm_motor_driver, w);
-	if (ret != 0) {
-		return UNABLE_TO_SET_PWM_CHNL1;
+#pragma region InternalFunctions// declarations of internal functions
+// They shouldn't be moved to .h file, their scope is only for .c file
+static void enc_callback_ch0(const struct device *dev, struct gpio_callback *cb, uint32_t pins);
+#if (CONFIG_SUPPORTED_CHANNEL_NUMBER > 1)
+static void enc_callback_ch1(const struct device *dev, struct gpio_callback *cb, uint32_t pins);
+#endif
+static int speed_pwm_set(uint32_t value);
+static void update_speed_and_position_continuus(struct k_work *work);
+static void continuus_calculation_timer_handler(struct k_timer *dummy);
+static void enc_callback(enum ChannelNumber chnl);
+#pragma endregion InternalFunctions
+
+/// Unit conversion defines
+#define MIN_TO_MS 60000
+#define RPM_TO_MRPM 1000
+
+/// CONTROL MODE - whether speed or position is controlled
+static enum ControlModes control_mode = SPEED;
+/// encoder timer - timer is common for both channels
+struct k_timer continuus_calculation_timer;
```

Comment:
everything (count_timer, continuus_calculation_timer) turned into static

Reply:
continuous_calculation_timer cannot be static, for some reason it breaks:

```
../src/driver.c:189:16: error: non-static declaration of 'continuous_calculation_timer' follows static declaration
  189 | K_TIMER_DEFINE(continuous_calculation_timer, continuous_calculation_timer_handler, NULL);
```

So it will be left as non-static